flagging_site/admin.py

```diff
@@ -0,0 +1,100 @@
+import os
+
+from flask import Flask
+from flask import redirect
+from flask import request
+from flask import Response
+from flask_admin import Admin
+from flask_admin import BaseView
+from flask_admin import expose
+from flask_admin.contrib import sqla
+from werkzeug.exceptions import HTTPException
+
+from .auth import AuthException
+from .auth import basic_auth
+from .data import db
+
+
+admin = Admin(template_mode='bootstrap3')
+
+def init_admin(app: Flask):
+    with app.app_context():
+        # Register /admin
+        admin.init_app(app)
+
+        # Register /admin sub-views
+        from .data.cyano_overrides import CyanoOverridesModelView
+        admin.add_view(CyanoOverridesModelView(db.session))
+        
+        admin.add_view(LogoutView(name="Logout"))
+
+
+# Adapted from https://computableverse.com/blog/flask-admin-using-basicauth
+class AdminModelView(sqla.ModelView):
+    """
+    Extension of SQLAlchemy ModelView that requires BasicAuth authentication,
+    and shows all columns in the form (including primary keys).
+    """
+
+    def __init__(self, model, *args, **kwargs):
+        # Show all columns in form
+        self.column_list = [c.key for c in model.__table__.columns]
+        self.form_columns = self.column_list
+
+        super().__init__(model, *args, **kwargs)
+
+    def is_accessible(self):
+        """
+        Protect admin pages with basic_auth.
+        If logged out and current page is /admin/, then ask for credentials.
+        Otherwise, raises HTTP 401 error and redirects user to /admin/ on the
+        frontend (redirecting with HTTP redirect causes user to always be
+        redirected to /admin/ even after logging in).
+
+        We redirect to /admin/ because our logout method only works if the path to
+        /logout is the same as the path to where we put in our credentials. So if
+        we put in credentials at /admin/cyanooverride, then we would need to logout
+        at /admin/cyanooverride/logout, which would be difficult to arrange. Instead,
+        we always redirect to /admin/ to put in credentials, and then logout at
+        /admin/logout.
+        """
+        if not basic_auth.authenticate():
+            if '/admin/' == request.path:
+                raise AuthException('Not authenticated. Refresh the page.')
```

Comment:
This is good. As a note: eventually we will need to add actual error handlers for stuff like this so the website doesn't completely crash on users.

I added an issue here relating to it: #81. You don't have to be the one who takes that issue mind you.

Reply:
I'm not sure it's worth putting in these error handlers for the /admin errors. For example, it won't show "Not authenticated. Refresh the page" unless the user goes to /admin and declines the prompt for credentials, and the other times when we raise a 401 error it should immediately redirect on the frontend to a more user-friendly page.